flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/changelog/StateChangelogClientTest.java

```diff
@@ -0,0 +1,158 @@
+        Map<Integer, List<byte[]>> appendsByKeyGroup = generateAppends(kgRange, 10, 20);
+
+        try (StateChangelogClient<?> client = clientType.create(this);
+                StateChangelogWriter<?> writer = client.createWriter(new OperatorID(), kgRange)) {
+            SequenceNumber from = writer.lastAppendedSqn();
+            appendsByKeyGroup.forEach(
+                    (group, appends) -> appends.forEach(bytes -> writer.append(group, bytes)));
+
+            StateChangelogHandle<?> handle = writer.persist(from).get();
+
+            assertByteMapsEqual(appendsByKeyGroup, extract(handle));
+        }
+    }
+
+    private void assertByteMapsEqual(
+            Map<Integer, List<byte[]>> expected, Map<Integer, List<byte[]>> actual) {
+        assertEquals(expected.size(), actual.size());
+        for (Map.Entry<Integer, List<byte[]>> e : expected.entrySet()) {
+            List<byte[]> expectedList = e.getValue();
+            List<byte[]> actualList = actual.get(e.getKey());
+            Iterator<byte[]> ite = expectedList.iterator(), ale = actualList.iterator();
+            while (ite.hasNext() && ale.hasNext()) {
+                assertArrayEquals(ite.next(), ale.next());
+            }
+            assertFalse(ite.hasNext());
+            assertFalse(ale.hasNext());
+        }
+    }
+
+    private Map<Integer, List<byte[]>> extract(StateChangelogHandle<?> handle) throws Exception {
+        Map<Integer, List<byte[]>> changes = new HashMap<>();
+        //noinspection unchecked
+        StateChangelogHandle<Object> objHandle = (StateChangelogHandle<Object>) handle;
+        try (CloseableIterator<StateChange> it = objHandle.getChanges(clientType.getContext())) {
+            while (it.hasNext()) {
+                StateChange change = it.next();
+                changes.computeIfAbsent(change.getKeyGroup(), k -> new ArrayList<>())
+                        .add(change.getChange());
+            }
+        }
+        return changes;
+    }
+
+    private Map<Integer, List<byte[]>> generateAppends(
+            KeyGroupRange kgRange, int keyLen, int appendsPerGroup) {
+        return stream(kgRange.spliterator(), false)
+                .collect(toMap(identity(), unused -> generateData(appendsPerGroup, keyLen)));
+    }
+
+    private List<byte[]> generateData(int numAppends, int keyLen) {
+        return Stream.generate(() -> randomBytes(keyLen))
+                .limit(numAppends)
+                .collect(Collectors.toList());
+    }
+
+    private byte[] randomBytes(int len) {
+        byte[] bytes = new byte[len];
+        random.nextBytes(bytes);
+        return bytes;
+    }
+
+    private enum LogClientType {
+        MEMORY {
+            @Override
+            public StateChangelogClient<?> create(StateChangelogClientTest logClientTest) {
+                return new InMemoryStateChangelogClient();
+            }
+
+            @Override
+            public StateChangeFormat getContext() {
```

Comment:
"StateChangeFormat" is missing?

Reply:
This was a leftover after splitting the changes into two PRs, sorry about that.